src/oracle.py

```diff
@@ -0,0 +1,135 @@
+
+
+class Oracle:
+    _last_finalized_epoch: EpochNumber = 0
+
+    def __init__(self, web3: Web3, beacon_chain_client: BeaconChainClient):
+        self._w3 = web3
+        self._beacon_chain_client = beacon_chain_client
+
+        self.modules: List[OracleModule] = [
+            Accounting(self._w3, self._beacon_chain_client),
+            Ejector(self._w3, self._beacon_chain_client),
+        ]
+
+    def _fetch_beacon_specs(self):
+        _, self.slots_per_epoch, self.seconds_per_slot, _ = contracts.oracle.functions.getBeaconSpec().call()
+        # _, self.slots_per_epoch, self.seconds_per_slot, _ = contracts.validator_exit_bus.functions.getBeaconSpec().call()
+
+    def run_as_daemon(self):
+        while True:
+            pulse()
+            logger.info({'msg': 'Run cycle.'})
+            self.run_cycle()
+
+    def run_cycle(self):
+        self._fetch_beacon_specs()
+        epoch = self._fetch_next_finalized_epoch()
+        logger.info({'msg': 'Get finalized epoch.', 'value': epoch})
+
+        try:
+            self.run_once(epoch)
+        except KeyboardInterrupt as error:
+            logger.error({'msg': 'Key interrupt.', 'error': error})
+            raise KeyboardInterrupt from error
+        except NoSlotsFound as error:
+            logger.error({'msg': 'No slots for current epoch.', 'error': error})
+        except Exception as error:
+            logger.error({'msg': 'Unexpected exception.', 'error': error})
+            raise Exception from error
+
+    def _fetch_next_finalized_epoch(self) -> EpochNumber:
+        while True:
+            current_finalized_epoch = EpochNumber(int(self._beacon_chain_client.get_head_finality_checkpoints()['finalized']['epoch']))
+
+            if current_finalized_epoch > self._last_finalized_epoch:
+                self._last_finalized_epoch = current_finalized_epoch
+
+                return current_finalized_epoch
+            else:
+                sleep_time = self.slots_per_epoch * self.seconds_per_slot / 4
+                logger.info({'msg': f'Waiting for next finalized epoch. Sleep for {sleep_time} sec.'})
+                sleep(sleep_time)
+
+    def run_once(self, epoch: Optional[EpochNumber] = None):
+        if epoch is None:
+            epoch = EpochNumber(int(self._beacon_chain_client.get_head_finality_checkpoints()['finalized']['epoch']))
+
+        FINALIZED_EPOCH_NUMBER.set(epoch)
+        slot = self._beacon_chain_client.get_first_slot_in_epoch(epoch, self.slots_per_epoch)
+        slot_number = SlotNumber(int(slot['message']['slot']))
+        SLOT_NUMBER.set(slot_number)
+        block_hash = slot['message']['body']['execution_payload']['block_hash']
+
+        logger.info({
+            'msg': 'Fetch first proposed slot in epoch.',
+            'epoch': epoch,
+            'slot': slot_number,
+            'block_hash': block_hash,
+        })
+
+        for module in self.modules:
+            try:
+                module.run_module(slot_number, block_hash)
```

Comment:
Is it ok, that every time when our code runs, "the accounting module" works first, and after the "ejector module" works. So, In the case when the Accounting module throws an exception, it means that the "ejector" module won't do its work every time. Also, if the oracle has more "modules" or workers themselves, that pattern will add affect cascade interrupting behavior all application.

Reply:
I thought those "modules" or "workers" had to work independently of each other.